core/src/main/java/org/jruby/ir/IRBuilder.java

```diff
@@ -3755,15 +3756,21 @@ public Operand buildReturn(ReturnNode returnNode) {
         Operand retVal = build(returnNode.getValueNode());
 
         if (scope instanceof IRClosure) {
-            // Closures return behavior has several cases (which depend on runtime state):
-            // 1. closure in method (return). !method (error) except if in define_method (return)
-            // 2. lambda (return) [dynamic]  // FIXME: I believe ->() can be static and omit LJE check.
-            // 3. migrated closure (LJE) [dynamic]
-            // 4. eval/for (return) [static]
-            boolean definedWithinMethod = scope.getNearestMethod() != null;
-            if (!(scope instanceof IREvalScript) && !(scope instanceof IRFor)) addInstr(new CheckForLJEInstr(definedWithinMethod));
-            addInstr(new NonlocalReturnInstr(retVal,
-                    definedWithinMethod ? scope.getNearestMethod().getName() : manager.runtime.newSymbol("--none--")));
+            if (scope.isWithinEND()) {
+                // ENDs do not allow returns
+                addInstr(new ThrowExceptionInstr(IRException.RETURN_LocalJumpError));
+            } else {
```

Comment:
MRI doesn't appear to statically determine this, or else they just immediately bail out without propagating an exception. Probably doesn't matter that we actually raise? Maybe better?

```
[] ~/projects/jruby $ rvm ruby-2.5.3 do ruby -e 'END { begin; return; rescue Exception => e; p e; else; puts "not caught"; end }'
-e: unexpected return

[] ~/projects/jruby $ jruby -e 'END { begin; return; rescue Exception => e; p e; else; puts "not caught"; end }'
#<LocalJumpError: unexpected return>
```

Reply:
yeah hmm.  I did not consider the ability to catch. I do actually consider how we do it to be more correct but of course we emulate them...Even if we change this I would prefer to follow up on this in another PR/commit later.  This is better than our escape crash.  I also am not very concerned about this behavioral difference?

Reply:
Sounds find to me.